src/workflow/logic.py

```diff
@@ -0,0 +1,116 @@
+from submission import models as submissions_models
+from proofing import models as proofing_models
+from production import models as production_models
+from journal import models as journal_models
+from core import models as core_models
+from utils.shared import clear_cache
+
+
+def reopen_proofing_assignments(article):
+    """
+    Finds and marks proofing assignments as open for a given article.
+    :param article: Article object
+    :return: None
+    """
+    for assignment in proofing_models.ProofingAssignment.objects.filter(
+        article=article,
+    ):
+        assignment.completed = None
+        assignment.save()
+
+
+def reopen_production_assignments(article):
+    """
+    Finds and marks production assignments as open for a given article.
+    :param article: Article objects
+    :return: None
+    """
+    for assignment in production_models.ProductionAssignment.objects.filter(
+        article=article,
+    ):
+        assignment.closed = None
+        assignment.accepted_by_manager = None
+        assignment.save()
+
+
+def reset_prepub_checklist(article):
+    """
+    Resets all prepub checklist items for a given article.
+    :param article: Article object
+    :return: None
+    """
+    for item in journal_models.PrePublicationChecklistItem.objects.filter(
+        article=article,
+    ):
+        item.completed = False
+        item.completed_by = None
+        item.completed_on = None
+        item.save()
+
+
+def drop_workflow_log_entries(article, stages_to_process):
+    """
+    Finds and deletes workflow log entries for stages being processed.
+    :param article: Article object
+    :param stages_to_process: A List of stage names
+    :return: None
+    """
+    del stages_to_process[-1]
```

Comment:
I don't think we want to mutate the list we've been passed.

Reply:
> I don't think we want to mutate the list we've been passed.

Maybe something like this would be succinct and readable?:

`*stages_to_delete, new_current_stage = stages_to_process`